GnuR modification: consider precompiled library functions called at the toplevel for JIT

vogr opened the pull request:
Fix for #1046

Corresponding PR in custom-r: https://github.com/reactorlabs/gnur/pull/10

vogr committed:
GnuR with modification to consider for RIR compilation functions
precompiled to BCODESXP in libraries.

correctly compute the jit score of bytecode compiled functions

o- committed:
not sure about this "fix"

o- committed:
workaround

o- committed:
fix fake cons cell leaking

o- committed:
fix memory leak of temp names array

o- committed:
temp fix for remove not clearing the bindings cache

o- committed:
remove broken ColonCastRHS constantfolding

o- committed:
update gnur

o- committed:
more robust fake cons-cells

o- committed:
missing protect

o- committed:
block more builtins

o- committed:
Revert "temp fix for remove not clearing the bindings cache"

This reverts commit 7233463ca2fd462b41e867f6d59bf145bd73b804.

o- committed:
properly fix binding cache when variable is removed

o- committed:
introduce PIR_OPT_LEVEL

to ensure these CI runs terminate eventually

o- committed:
Revert "Revert "temp fix for remove not clearing the bindings cache""

This reverts commit 8e70ea04d41e8d851a619aa75eb74b669d8bde76.

o- committed:
this should not happen anymore

o- committed:
see what breaks

o- committed:
fix pidigits regression

need some support for $

o- committed:
Revert "this should not happen anymore"

This reverts commit d8b01c0d9f57af7c7d9814ef067f9d4829ae178a.

o- committed:
disable these asserts

o- committed:
more bugfixes

o- committed:
fix class<-

o- committed:
make tests pass for all opt levels

o- committed:
don't promise wrap constant args

o- committed:
fix typechecks for non-attribute objects

o- committed:
prevent double evaluating $ lhs

o- committed:
fix some regressions

o- committed:
missing cf

o- committed:
improve $ support

o- committed:
trying a different strategy for attrib/obj types

o- committed:
protect fix

o- committed:
more debugging

o- committed:
more protect

o- committed:
debugging

o- committed:
fix for loop for factors

o- committed:
this should fix no-segv error

o- committed:
fighting timeouts

o- committed:
small fixes

o- committed:
fix another issue with rm/remove(...)

o- committed:
fix for uncached envs

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 3.6h
* are-we-fast-r suite regressed by 0.97
* real_thing suite improved by 1.1
* Bounce_nonames_simple regressed by 0.9
* pidigits regressed by 0.79
* mandelbrot_noout regressed by 0.95
* emptyFor regressed by 0.92
* vectorWhile regressed by 0.93
* listWhile regressed by 0.94
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1441560629&job_ids[]=1432940244&selection=all

o- committed:
speeding up tests

o- committed:
fix for loop sequence conversion in pir

o- committed:
fix switch and for loops

switch was broken because of using as.integer builtin instead of what
switch actually does (silently converting to int).

isFactor was not implemented in native

for sequences were broken in native as it is not correct to modify them
in-place there.

o- committed:
fix ci

o- committed:
fix cppcheck

o- committed:
fast isFactor check

o- committed:
yet again binding caches...

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_features_3 took 3.71h
* real_thing suite improved by 1.08
* Bounce_nonames_simple regressed by 0.95
* convolution_v regressed by 0.88
* convolution_slow regressed by 0.92
* pidigits regressed by 0.83
* fasta_2 regressed by 0.94
* nbody_naive_2 regressed by 0.94
* mandelbrot_noout regressed by 0.95
* emptyFor regressed by 0.92
* listWhile regressed by 0.94
* listFor regressed by 0.94
* Overall benchmarks improved by 1.01

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1455062241&job_ids[]=1432940244&selection=all

o- committed:
lessen regression due to slower isFactor

o- committed:
faster tests

o- committed:
fixing more tests

o- committed:
random guess

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 3.24h
* real_thing suite improved by 1.03
* convolution_v regressed by 0.9
* convolution_slow regressed by 0.92
* pidigits regressed by 0.78
* mandelbrot_noout regressed by 0.95
* scalarFor regressed by 0.93
* vectorWhile regressed by 0.93
* listWhile regressed by 0.94
* emptyFor regressed by 0.94
* listFor regressed by 0.94
* Overall benchmarks regressed by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=1462279867&job_ids[]=1450318169&selection=all
